packages/teleterm/src/ui/commandLauncher.ts

```diff
@@ -15,15 +15,81 @@ limitations under the License.
 */
 
 import { IAppContext } from 'teleterm/ui/types';
+import { routing } from 'teleterm/ui/uri';
+import { tsh } from 'teleterm/ui/services/clusters/types';
 
 const commands = {
+  // For handling the "Connect" button next to a server.
   ssh: {
     displayName: '',
     description: '',
     run(ctx: IAppContext, args: { serverUri: string }) {
       ctx.modalsService.openProxySshDialog(args.serverUri);
     },
   },
+  // For handling "tsh ssh" executed from the command bar.
+  'tsh-ssh': {
+    displayName: '',
+    description: '',
+    run(ctx: IAppContext, args: { loginHost: string; clusterUri: string }) {
+      const { loginHost, clusterUri } = args;
+      const documentsService =
+        ctx.workspacesService.getWorkspaceDocumentService(clusterUri);
+      let login: string | undefined, host: string;
+      const parts = loginHost.split('@');
+
+      if (parts.length > 1) {
+        host = parts.pop();
+        // If someone types in `foo@bar@baz` as input here, `parts` will have more than two
+        // elements. `foo@bar` is probably not a valid login, but we don't want to lose that
+        // input here.
+        //
+        // In any case, we're just repeating what `tsh ssh` is doing with inputs like these.
+        login = parts.join('@');
+      } else {
+        host = parts[0];
+      }
+
+      // TODO(ravicious): Handle finding host by more than just a name.
+      // Basically we have to replicate tsh ssh behavior here.
+      const servers = ctx.clustersService.searchServers(clusterUri, {
+        search: host,
+        searchableProps: ['hostname'],
+      });
+      let server: tsh.Server | undefined;
+
+      if (servers.length === 1) {
+        server = servers[0];
+      } else if (servers.length > 1) {
+        // TODO(ravicious): Handle ambiguous host name. See `onSSH` in `tool/tsh/tsh.go`.
+        console.error('Ambiguous host');
+      }
+
+      let serverUri: string, serverHostname: string;
+
+      if (server) {
+        serverUri = server.uri;
+        serverHostname = server.hostname;
+      } else {
+        // If we can't find a server by the given hostname, we still want to create a document to
+        // handle the error further down the line.
+        const { rootClusterId, leafClusterId } =
+          routing.parseClusterUri(clusterUri).params;
+        serverUri = routing.getServerUri({
+          serverId: host,
+          rootClusterId,
+          leafClusterId,
+        });
+        serverHostname = host;
+      }
+      // TODO(ravicious): Handle failure due to incorrect host name.
```

Comment:
We haven't talked much about error handling for the command bar in general. Right now if the pty process exits, we just close the tab. I think we should just not close it if the exit code is a non-success code.

Reply:
Agree